### DISCUSSION TITLE
How to draw images without them being cached

### DISCUSSION THREAD

**Discussion by marc2332** [rust-skia/rust-skia#949] [Q&A] [answered]
Hey!

I have been rendering images and I think they have been leaking memory because they get cached. Here is some code of how I do it:
```rs
let pic = Image::from_encoded(unsafe { Data::new_bytes(bytes) });
if let Some(pic) = pic {
    let mut paint = Paint::default();
    paint.set_anti_alias(true);
    canvas.draw_image_nine(
        pic,
        IRect::new(0, 0, 0, 0),
        Rect::new(area.min_x(), area.min_y(), area.max_x(), area.max_y()),
        FilterMode::Last,
        Some(&paint),
    );
}
```

I have seen some mentions of a `CachingHint` but I can't figure out how to pass it

**Comment by marc2332** [comment]
hey @tronical, sorry for pinging but I thought you might have not seen this

**Reply by tronical** [reply]
Hm, I haven't seen this, but maybe I also haven't looked ;). I'm reasonably certain that the `canvas` will keep a reference to the `Image` until at least `flush()` is called on the `DirectContext` (if you're using gpu rendering).

That said, just to be sure, you also see this if you use `new_copy` instead of `Data::new_bytes`?

(I think the CachingHint works in the other direction - when downloading a texture to the CPU and that transfer can be cached)

**Reply by marc2332** [reply]
omg, I didn't mean to ping you ! I wanted to ping @pragmatrix, I hate GitHub's autocomplete sometimes... Sorry!

**Reply by marc2332** [reply]
Let me try with `new_copy` though

**Reply by marc2332** [reply]
It seems to behave exactly the same with `new_copy`

**Reply by pragmatrix** [reply, marked as answer]
Hey Marc, I am not sure, but may be the images get cached in the resource cache, the `skia_safe::graphics::resource_cache*` functions may help, and there you can set limits, too.

**Reply by marc2332** [reply]
That seems to be it! Thanks @pragmatrix 😄 Good afternoon :v:
